### PR TITLE
Prepare to remove project from plugin context (6 of X)

### PR DESCRIPTION
**Pull request by tobous**
Looks like a lot of code, but most of the changes are minor refactorings or the addition of simple data holder classes and their factories.

The PR should by reviewed by going through the commits. Each commit features a commit message giving more details about the content.

### CODE REVIEW

File: intellij/src/saros/intellij/context/SharedIDEContext.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+import saros.intellij.eventhandler.IApplicationEventHandler.ApplicationEventHandlerType;
+import saros.intellij.eventhandler.IProjectEventHandler.ProjectEventHandlerType;
+import saros.intellij.eventhandler.IStartableDisableableHandler;
+import saros.intellij.eventhandler.ProjectEventHandlers;
+import saros.intellij.eventhandler.ProjectEventHandlersFactory;
+import saros.repackaged.picocontainer.Disposable;
+
+/**
+ * Singleton class representing the shared IDE context. It contains the session components dependent
+ * on IDE specific resources. This main consist of the shared project references and the
```
Comment (m273d15) [inline intellij/src/saros/intellij/context/SharedIDEContext.java]: Typo `main` -> `mainly`

Comment (tobous) [reply intellij/src/saros/intellij/context/SharedIDEContext.java]: Done.

**Comment by tobous** [comment]
Fixed minor typo in javadoc. Rebased on current master without any conflicts.

**Comment by tobous** [comment]
Add missing `NotNull` annotations.

**Comment by m273d15** [comment]
> Its interesting to me that Codacy sees the import of Project as unnecessary in ApplicationEventHandlers and ProjectEventHandlers. While it is not used in the code itself, it is used in the class javadoc (which is why neither the IDE nor google java format complained). This seems like a false-positive to me.

@tobous It seems as if our formatter enforces a line-break within the `{@link ...}` block. Therefore PMD (the analyser that throws this violation) has the [following issue](https://github.com/pmd/pmd/issues/1639).

**Comment by tobous** [comment]
Rebased onto current master without any interaction.

**Pull request merged**